### PR TITLE
[HIVEMALL-17] Support SLIM

### PR DESCRIPTION
**Pull request by nzw0301**
## What changes were proposed in this pull request?

- Add new UDTF: `train_slim`
- Add new evaluation UDAF for ranking evaluation measures: `HitRate` and `ARHR`
- Fix typos

## What type of PR is it?

Improvement

## What is the Jira issue?

https://issues.apache.org/jira/browse/HIVEMALL-17

## How was this patch tested?

- Add `recommend/SlimUDTFTest`
- Manual tests for MovieLens1m on EMR
  - please check the evaluation value in the `movielens_slim.md`

## How to use this feature?

- Please see markdown file: `movielens_slim.md` in this PR.

## Checklist

- [x] Did you apply source code formatter, i.e., `mvn formatter:format`, for your commit?

### CODE REVIEW

File: core/src/main/java/hivemall/recommend/SlimUDTF.java [hunk 1]
```diff
@@ -144,11 +208,76 @@ public void process(Object[] args) throws HiveException {
...
+        }
+        srcBuf.clear();
     }
 
     @Override
     public void close() throws HiveException {
+
+        runIterativeTraining();
+
         int numItem = Math.max(this.W.numRows(), this.W.numColumns());
```
Comment (myui) [inline core/src/main/java/hivemall/recommend/SlimUDTF.java]: Please add the following method in `DoKMatrix` and use it.

```java
    public void eachNonZeroCell(@Nonnull final VectorProcedure procedure) {
        if (nnz == 0) {
            return;
        }
        final IMapIterator itor = elements.entries();
        while (itor.next() != -1) {
            long k = itor.getKey();
            int row = Primitives.getHigh(k);
            int col = Primitives.getLow(k);
            double value = itor.getValue();
            procedure.apply(row, col, value);
        }
    }

    @Override
    public RowMajorMatrix toRowMajorMatrix() {
```

```java
public abstract class VectorProcedure {
    ...
    public void apply(@Nonnegative int row, @Nonnegative int col, double value) {}

}
```

File: core/src/main/java/hivemall/math/matrix/sparse/DoKMatrix.java [hunk 2]
```diff
@@ -309,6 +309,20 @@ public void eachNonZeroInColumn(@Nonnegative final int col,
...
+        if (nnz == 0) {
+            return;
+        }
+        final IMapIterator itor = elements.entries();
+        while (itor.next() != -1) {
+            long k = itor.getKey();
+            int row = Primitives.getHigh(k);
+            int col = Primitives.getLow(k);
+            double value = itor.getValue();
+            procedure.apply(row, col, value);
```
Comment (myui) [inline core/src/main/java/hivemall/math/matrix/sparse/DoKMatrix.java]: `VectorProcedure#apply(@Nonnegative int row, @Nonnegative int col, double value)` should also be introduced.

Comment (nzw0301) [reply core/src/main/java/hivemall/math/matrix/sparse/DoKMatrix.java]: Oop, Thanks!

File: core/src/main/java/hivemall/recommend/SlimUDTF.java [hunk 3]
```diff
@@ -217,13 +347,48 @@ private void train(int i, Map<?, ?> Ri, Map<?, ?> topKRatesOfI, int j, Map<?, ?>
...
+//
+//        gradSum /= N;
+//        rateSum /= N;
+//        errs /= N;
+//
+//        this.loss += errs;
+//        this.W.unsafeSet(i, j, getUpdateTerm(gradSum, rateSum));
+//    }
+
+    private double getUpdateTerm(double gradSum, double rateSum){
```
[the remark this answers is not in the document]
Comment (nzw0301) [reply core/src/main/java/hivemall/recommend/SlimUDTF.java]: Sorry, I missed an error related this definition.
This function refers class variables `l1` and `l2`, so I will change it to `getUpdateTerm(final double gradSum, final double rateSum, final double l1, final double l2)`.

File: resources/ddl/define-all.hive [hunk 4]
```diff
@@ -694,6 +694,12 @@ create temporary function rf_ensemble as 'hivemall.smile.tools.RandomForestEnsem
 drop temporary function if exists guess_attribute_types;
 create temporary function guess_attribute_types as 'hivemall.smile.tools.GuessAttributesUDF';
 
+--------------------
+-- recommendation --
+--------------------
+drop temporary function if exists train_slim;
+create temporary function train_slim as 'hivemall.recommend.SlimUDTF';
```
Comment (myui) [inline resources/ddl/define-all.hive]: define `train_slim` in the other DDL files. grep `train_mf` for reference.

File: docs/gitbook/recommend/movielens_cf.md [hunk 5]
```diff
@@ -61,7 +61,8 @@ partial_result as ( -- launch DIMSUM in a MapReduce fashion
     movie_features f
   left outer join movie_magnitude m
 ),
-similarity as ( -- reduce (i.e., sum up) mappers' partial results
+similarity as (
+    -- reduce (i.e., sum up) mappers' partial results
```
Comment (myui) [inline docs/gitbook/recommend/movielens_cf.md]: Is it possible to use `'` is comments in Hive v2.3.0?

Comment (nzw0301) [reply docs/gitbook/recommend/movielens_cf.md]: Yes.

Comment (nzw0301) [reply docs/gitbook/recommend/movielens_cf.md]: If the number of `'` is even in comment, `'` can be used in the same line:

```
> select (1 --'sample'
> );
```

**Comment by nzw0301** [comment]
After a few checks, I'll remove [WIP] tag in this PR.

- whether hive input KNN data can be read from tem file
- whether loss function correctly decreases

File: core/src/main/java/hivemall/utils/collections/maps/Int2DoubleOpenHashTable.java [hunk 6]
```diff
@@ -0,0 +1,417 @@
...
+            return true;
+        }
+        if (stat == REMOVED && _keys[index] == key) {
+            return true;
+        }
+        return false;
+    }
+
+    /** @return expanded or not */
+    protected boolean preAddEntry(int index) {
```
Comment (nzw0301) [inline core/src/main/java/hivemall/utils/collections/maps/Int2DoubleOpenHashTable.java]: `index` is never used in this function. Other `int*OpenHashMap`s functions are similar

Comment (myui) [reply core/src/main/java/hivemall/utils/collections/maps/Int2DoubleOpenHashTable.java]: It's for overriding and thus okey as it is.

Comment (nzw0301) [reply core/src/main/java/hivemall/utils/collections/maps/Int2DoubleOpenHashTable.java]: I see. Thanks.

File: core/src/main/java/hivemall/recommend/SlimUDTF.java [hunk 7]
```diff
@@ -0,0 +1,625 @@
...
+        int numIterations = 40;
+        boolean conversionCheck = true;
+        double cv_rate = 0.005d;
+
+        if (argOIs.length >= 6) {
+            String rawArgs = HiveUtils.getConstString(argOIs[5]);
+            cl = parseOptions(rawArgs);
+
+            l1 = Primitives.parseDouble(cl.getOptionValue("l1"), l1);
+            if (l1 < 0.d || l1 > 1.d) {
```
Comment (myui) [inline core/src/main/java/hivemall/recommend/SlimUDTF.java]: `if (l1 < 0.d)` is enough.

File: core/src/main/java/hivemall/recommend/SlimUDTF.java [hunk 8]
```diff
@@ -0,0 +1,625 @@
...
+                throw new UDFArgumentException("Argument `double l1` must be within [0., 1.]: "
+                        + l1);
+            }
+
+            l2 = Primitives.parseDouble(cl.getOptionValue("l2"), l2);
+            if (l2 < 0.d || l2 > 1.d) {
```
Comment (myui) [inline core/src/main/java/hivemall/recommend/SlimUDTF.java]: ` if (l2 < 0.d) {` is enough.

**Comment by nzw0301** [comment]
@myui @takuti If you have a time, could you review this PR?

File: core/src/main/java/hivemall/recommend/SlimUDTF.java [hunk 9]
```diff
@@ -0,0 +1,636 @@
...
+    private ConversionState cvState;
+    private long observedTrainingExamples;
+
+    public SlimUDTF() {}
+
+    @Override
+    public StructObjectInspector initialize(ObjectInspector[] argOIs) throws UDFArgumentException {
+        final int numArgs = argOIs.length;
+
+        if (numArgs == 1 && HiveUtils.isStringOI(argOIs[0])) {
```
Comment (nzw0301) [inline core/src/main/java/hivemall/recommend/SlimUDTF.java]: I add this a few line to show the slim explanation without other arguments. But it may not be conventional way to show `help` for hivemall.

**Comment by myui** [comment]
@nzw0301 could you rebase with the master branch?

**Comment by nzw0301** [comment]
@myui ok, please give me a time.

**Comment by nzw0301** [comment]
@myui done.

File: core/src/main/java/hivemall/evaluation/GradedResponsesMeasures.java [hunk 10]
```diff
@@ -54,4 +54,23 @@ public static double DCG(final List<Double> topRelScoreList, final int recommend
...
+
+    /**
+     * Computes Reciprocal HitRank (RHR)
+     *
+     * @param recommendList predicted item list order by score
+     * @param truthList gruond truth item list order by rank
+     * @param recommendSize the number of positive items
+     * @return RHR
+     */
+    public static double RHR(final List<?> recommendList, final List<?> truthList,
```
Comment (myui) [inline core/src/main/java/hivemall/evaluation/GradedResponsesMeasures.java]: @nzw0301 (cc: @takuti )

This is not graded response but binary response.

`ReciprocalRank` is already implemented in `BinaryResponseMeasures` and  `MRRUDAF` implements [Mean reciprocal rank](https://en.wikipedia.org/wiki/Mean_reciprocal_rank).

ARHRUDAF is duplicate to MRRUDAF and MRR is more generic name of ARHR.

[Average reciprocal hit-rate is also referred to as the mean reciprocal rank (MRR)](https://books.google.co.jp/books?id=GKjWCwAAQBAJ&pg=PA246&lpg=PA246&dq=%22average+reciprocal+hit-rate%22+MRR&source=bl&ots=Tq_wVfI-QU&sig=Hbyh-R1Sm7M26pZbHPjuTfyXXqI&hl=ja&sa=X&ved=0ahUKEwie1MD0prPWAhWMW7wKHVHzD88Q6AEIJzAA#v=onepage&q=%22average%20reciprocal%20hit-rate%22%20MRR&f=false)

File: core/src/main/java/hivemall/evaluation/HitRateUDAF.java [hunk 11]
```diff
@@ -0,0 +1,261 @@
...
+            List<?> truthList = truthListOI.getList(parameters[1]);
+            if (truthList == null) {
+                return;
+            }
+
+            int recommendSize = recommendList.size();
+            if (parameters.length == 3) {
+                recommendSize = recommendSizeOI.get(parameters[2]);
+            }
+            if (recommendSize < 0 || recommendSize > recommendList.size()) {
```
Comment (myui) [inline core/src/main/java/hivemall/evaluation/HitRateUDAF.java]: Hit rate should accept `recommendSize > recommendList.size()`.

**Comment by myui** [comment]
I'll create another PR `[HIVEMALL-17-2]` based on this PR.

File: core/src/test/java/hivemall/recommend/SlimUDTFTest.java [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+        ObjectInspector itemJRatesOI = ObjectInspectorFactory.getStandardMapObjectInspector(
+            PrimitiveObjectInspectorFactory.javaIntObjectInspector,
+            PrimitiveObjectInspectorFactory.javaDoubleObjectInspector);
+        ObjectInspector knnOfIOI = ObjectInspectorFactory.getStandardMapObjectInspector(
+            PrimitiveObjectInspectorFactory.javaIntObjectInspector,
+            ObjectInspectorFactory.getStandardMapObjectInspector(
+                PrimitiveObjectInspectorFactory.javaIntObjectInspector,
+                PrimitiveObjectInspectorFactory.javaDoubleObjectInspector));
+        ObjectInspector argumentOI = ObjectInspectorUtils.getConstantObjectInspector(
+            PrimitiveObjectInspectorFactory.javaStringObjectInspector, "-l1 2.");
```
Comment (myui) [inline core/src/test/java/hivemall/recommend/SlimUDTFTest.java]: `2.` is valid and parsed as `2.0`.

File: core/src/main/java/hivemall/recommend/SlimUDTF.java [hunk 13]
```diff
@@ -0,0 +1,636 @@
...
+        if (this.dataMatrix == null) {
+            this.dataMatrix = new DoKMatrix();
+        }
+
+        if (this.weightMatrix == null) {
+            this.weightMatrix = new DoKMatrix();
+        }
+
+        int itemI = PrimitiveObjectInspectorUtils.getInt(args[0], itemIOI);
+        Int2FloatOpenHashTable ri = map2Int2FloatOpenHashTable(this.riOI.getMap(args[1]), riKeyOI,
```
Comment (myui) [inline core/src/main/java/hivemall/recommend/SlimUDTF.java]: `ri` and `knnItems` can be reused for each item `i`.

```
process(i, Map<user, rating> Ri, Map<user, List<item, rating>> KNNi, j, Map<user, rating> Rj) {// for each NN (j) of base item i
  if(prev_i != i) {
     runIterations();
     this.prev_i = i;
     A.clear(); K.clear();
  }
```
